src/shims/io.rs

```diff
@@ -0,0 +1,192 @@
+            this.get_handle_and(fd, |handle| Ok(handle.flag), -1)
+        } else {
+            throw_unsup_format!("Unsupported command {:#x}", cmd);
+        }
+    }
+
+    fn close(&mut self, fd_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+
+        this.remove_handle_and(
+            fd,
+            |handle, this| this.consume_result::<i32>(handle.file.sync_all().map(|_| 0), -1),
+            -1,
+        )
+    }
+
+    fn read(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        buf_op: OpTy<'tcx, Tag>,
+        count_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i64> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let tcx = &{ this.tcx.tcx };
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
+
+        // Remove the file handle to avoid borrowing issues
+        this.remove_handle_and(
+            fd,
+            |mut handle, this| {
+                let bytes = handle
+                    .file
+                    .read(this.memory_mut().get_mut(buf.alloc_id)?.get_bytes_mut(
+                        tcx,
+                        buf,
+                        Size::from_bytes(count),
+                    )?)
+                    .map(|bytes| bytes as i64);
+                // Reinsert the file handle
+                this.machine.file_handler.handles.insert(fd, handle);
+                this.consume_result::<i64>(bytes, -1)
+            },
+            -1,
+        )
+    }
+
+    fn get_handle_and<F, T>(&mut self, fd: i32, f: F, t: T) -> InterpResult<'tcx, T>
+    where
+        F: Fn(&FileHandle) -> InterpResult<'tcx, T>,
+    {
+        let this = self.eval_context_mut();
+        if let Some(handle) = this.machine.file_handler.handles.get(&fd) {
+            f(handle)
+        } else {
+            this.machine.last_error = this.eval_libc_i32("EBADF")? as u32;
+            Ok(t)
+        }
+    }
+
+    fn remove_handle_and<F, T>(&mut self, fd: i32, mut f: F, t: T) -> InterpResult<'tcx, T>
+    where
+        F: FnMut(FileHandle, &mut MiriEvalContext<'mir, 'tcx>) -> InterpResult<'tcx, T>,
+    {
+        let this = self.eval_context_mut();
+        if let Some(handle) = this.machine.file_handler.handles.remove(&fd) {
+            f(handle, this)
+        } else {
+            this.machine.last_error = this.eval_libc_i32("EBADF")? as u32;
+            Ok(t)
+        }
+    }
+
+    fn consume_result<T>(&mut self, result: std::io::Result<T>, t: T) -> InterpResult<'tcx, T> {
```

Comment:
is anything but `-1: i32` ever used for `t: T`? If not, remove the argument

Reply:
same for the above functions

Reply:
well the thing is that in some cases `-1i64` is returned. I  didn't manage to return both in the same function without passing it as an argument

Reply:
well the thing is that in some cases -1i64 is returned. I didn't manage to return both in the same function without passing it as an argument

Reply:
I would have thought all the C-stuff to use `c_int` which is `i32`. Where is `i64` needed?

Reply:
for `read` (and `write` eventually) but I was able to fix it using `From<i32>` :P